src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4

```diff
@@ -39,6 +39,24 @@ options { tokenVocab=JavaLanguageLexer; }
     private boolean isYieldStatement() {
         return _input.LT(1).getType() == JavaLanguageLexer.LITERAL_YIELD && switchBlockDepth > 0;
     }
+
+    static int fileCounter = 0;
+
+    /**
+     * We create a custom constructor so that we can clear the DFA
+     * states upon instantiation of JavaLanguageParser; this requires
+     * adding a dummy argument to constructor parameters.
+     *
+     * @param input the token stream to parse
+     * @param dummyArg dummy argument
+     */
+    public JavaLanguageParser(TokenStream input, boolean dummyArg) {
+        super(input);
+        _interp = new ParserATNSimulator(this, _ATN , _decisionToDFA, _sharedContextCache);
+        if (++fileCounter % 500 == 0) {
```

Comment:
Clear the DFA states for every 500 files parsed.

Reply:
@nmancus1 I'm ok with this constant, but what if we turn this into a parameter? Instead the  `dummyArg` there can be `dfaLimit` or something like that. It is even possible to add a property to the TreeWalker to make it configurable by the user.

Reply:
Done. Adding a limit of files to parse before clearing the dfa states allowed us to continue using JavaParser#parse in test mentioned at https://github.com/checkstyle/checkstyle/pull/10999#discussion_r776793760. Thanks for the idea!